wip: deployment doc (#435)
inveniosoftware/docs-invenio-rdm: docs/deploy/index.md

These charts would enable you to deploy your instance on top of
[Kubernetes](https://kubernetes.io) clusters too. If maintaining the actual
PaaS technology is not something you desire, you can make use of cloud
providers such as [Amazon Web Services](https://aws.amazon.com/),
[Google Cloud](https://cloud.google.com/), [Microsoft Azure](https://azure.microsoft.com/).

- Infrastructure as a Service (IaaS): using this model you will be in charge of
the stack from the operative system up. This is more commonly known as _virtual machines_,
meaning that you will be in charge of choosing the python version, the web server, etc.
At CERN we have been running Invenio v3 instances using [OpenStack](https://www.openstack.org)
to manage the VMs, and [Puppet](https://www.puppet.com) for IaC (Infrastructure as Code).
Other common IaC technologies are [Ansible](https://www.ansible.com) and [Terraform](https://www.terraform.io).

![Deployment models](../images/deployment-models.png)

You could also keep going down in the layers to get more and more control of
your deployment, for example, using bare metal. However, that is out of the
scope of this documentation.

> !!! warning "Docker compose is not recommended"

Reviewer: I do agree on "seeing it on the menu", as in visible to avoid mistakes. However, it does not really fit semantically in the higher level menu...

    The provided docker-compose files can give you a good idea on what you need
    to deploy, including the underlying services. However, the provided files
    are **NOT** meant to be used in production as-is: it requires knowledge
    and experience on how to run it reliable, safely stores files, understand
    potential performance limitations, etc. Use at your own risk.

## Services

When deploying InvenioRDM you can choose to deploy and maintain yourself the
services, or use third party providers. For example, the three cloud providers
mentioned above (AWS, GC, and Azure) can provide most of the services. If you

> choose to deploy them yourself you will need to take into account the following

Author: What if we make it more "scary"? :)
```suggestion
choose to deploy them yourself you will need to deep-dive and get experienced in the following topics:
```

topics:

- Persist your data, and enable periodic backups. This includes the relational
database, the search indices and the files.
- Queue persistence (RabbitMQ) to avoid loosing tasks in case the service fails.
- High availability, many of the services can be deployed redundantly. Note
that most of the cloud providers offer this option or have an established SLA.
- Secrets handling. Most of the services require credentials to connect. It is
a good practice to keep them in a secure place, for example [Ansible Vault](https://www.redhat.com/sysadmin/introduction-ansible-vault).

> 

Author: Shall we add a bullet point about Security?

If you choose to deploy them using PaaS, images for the services can be
found for example in [DockerHub](https://hub.docker.com). For more information
on how to deploy each service, refer to their official documentation.

## InvenioRDM Python application

Independently of your deployment model choice, you will need to run the Flask
Python application. However, the server bundled with Flask is not meant for
production usage. Therefore, you will need some sort of WSGI HTTP server.
Common choices are [Gunicorn](https://gunicorn.org) and [uWSGI](http://projects.unbit.it/uwsgi),

> The latter, is the one we use at CERN and it is bundled in the Dockerfile.

Author: Does it make sense to add a link to the other doc section about the RDM Docker images?

### Versioning

During the lifetime of your service you will develop new features specific to
your instance or need to upgrade to a newer version of InvenioRDM. This will
means that your application code will change and versioning it will help you
control which code is actually deployed. At CERN we use GitHub tags/release

> for each version (e.g. v1.0.0) plus a `latest` and `production` tag, these two

Author: Suggestion: I might be tempted to say to remove the `latest` and `production` tags, they can be dangerous if you don't know exactly what you are doing.

Reviewer: Agree

will point to the latest development and production release respectively.

If you are using container for the deployment, you can automate the image
build with for example [GitHub actions](https://github.com/features/actions)
(or any other CI tool). In addition, some PaaS platforms have the capabilities
to detect when a new image for a certain tag (e.g. `production`) has changed
and re-deploy it automatically.

## Monitoring

After your application is deployed and running, monitoring it at many level
will provide you the observability needed to be aware of what is happening
at all times.

Once you have configured where the different services output their logs
they can be aggregated and create dashboards to visualize them, for example in

> [Loki](https://grafana.com/oss/loki/) or [OpenSearch](https://opensearch.org/docs/latest/dashboards/index/).

Author: ```suggestion
[Loki](https://grafana.com/oss/loki/) or [OpenSearch Dashboards](https://opensearch.org/docs/latest/dashboards/index/).
```

In a IaaS deployment it might be useful to aggregate also the system logs to
monitor the resources consumption. This metrics can be aggregated using
[Prometheus](https://prometheus.io) and visualized in [Grafana](https://grafana.com).

In some cases it is possible to have application specific monitoring. For
example, using [Flower](https://flower.readthedocs.io/en/latest/) to monitor
the queues status. Moreover, many of these can be plugged in to Prometheus.
Closing that way the circle and providing a single point of view for your
monitoring and metrics.

> For the Python application [Sentry](https://sentry.io/welcome/) provides a

Author: We might want to add as an option to save logs in files, rotating.

great level of details (variables' values, stacktraces, etc.), which is
specially helpful when it comes to understanding errors that happen in

> production.

Author: ```suggestion
production. **We really recommend it!**
```

### Alerting

Once the monitoring is in place, the next steps is getting alerts when
certain metrics reach a configured threshold. In the Grafana
ecosystem, [OnCall](https://grafana.com/products/oncall). Alternatively, many
messaging apps can be configured with [webhooks to Grafana](https://grafana.com/docs/grafana/latest/alerting/manage-notifications/webhook-notifier/).

## Security

TODO: https://github.com/inveniosoftware/training/tree/v3.4/13-securing-your-instance/

## Load testing

> It might be interesting to understand what is the load that your system will sustain.

Author: ```suggestion
It might be interesting to understand what is the load that your system will sustain, to size your infrastructure accordingly.
```

That can be tested for example with [locust](https://locust.io) or [k6](https://k6.io).

## Troubleshooting

The first step to troubleshooting a problem would be to try to understand in
which layer or service of the system is the problem. That can be checked using

> the monitoring and dashboard technologies mentioned above. If the problem is in

Author: Add to check the logs files.
